test/testutilsgzip.cpp

```diff
@@ -48,6 +49,18 @@ private slots:
         bool ok = false;
         const QByteArray compressedData = Utils::Gzip::compress(data, 6, &ok);
         QVERIFY(ok);
+
+        ok = false;
+        QBuffer inBuff;
+        inBuff.setData(data);
+        inBuff.open(QIODevice::ReadOnly);
+        QBuffer outBuff;
+        outBuff.open(QIODevice::WriteOnly);
+        ok = Utils::Gzip::compress(inBuff, outBuff, 6);
+        inBuff.close();
+        outBuff.close();
+        QVERIFY(ok);
+        QCOMPARE(compressedData, outBuff.data());
```

Comment:
Please move it to its own function. Usually there is one-to-one mapping of private function in test class and the tested function.

Also you should thoroughly test it, for example:
* `Utils::Gzip::compress(QIODevice(), buf);`
* `Utils::Gzip::compress(buf, QIODevice());`
* `Utils::Gzip::compress(QIODevice(), QIODevice());`

The same applies to `decompress` too.

Reply:
@Chocobo1 
It seems that you are too biased about this. We are not writing a public library of utilities. Maybe it would be worth limiting the requirements? Even in public interfaces, the responsibility for the correctness of parameters is often left to the caller, so as not to bring matters to the point of absurdity, when a function devotes 90% of logic to checking parameters instead of useful work.

Reply:
>It seems that you are too biased about this.

Have you ever thought thoroughly of writing tests and their meanings? Or can you stop your bad habit of accusing?

>Maybe it would be worth limiting the requirements? 

In fact, I consider the suggested tests to be crucial, others I may not care. As you can see from qt docs, the underlying device won't be available all the time and therefore the function should handle it correctly. At the very least, the function should not crash and these tests ***will*** ensure it.
I've seen too many instances of the developer being too much of himself and imagine his code is bulletproof and won't crash.

Reply:
>Even in public interfaces, the responsibility for the correctness of parameters is often left to the caller, so as not to bring matters to the point of absurdity, when a function devotes 90% of logic to checking parameters instead of useful work.

Here is a counter example for you: https://github.com/qbittorrent/qBittorrent/blob/ad491652336e24f04be9c874acf5b338d96b492b/src/base/utils/gzip.cpp#L121-L122
Note that the tests didn't cover it...

Reply:
> can you stop your bad habit of accusing?

I'm not accusing anyone, and even more so I don't have such a habit.
>Have you ever thought thoroughly of writing tests and their meanings? 

Sure. They need to make sure that the function (still) works as expected. But isn't it obvious that you first need to decide on what is "works as expected" in each specific case? That's what I was talking about.
In this case, we could make an explicit separation of the input parameters into valid and invalid, and assign responsibility for this to the caller, and expect the function to work correctly only with valid parameters (so tests will follow only this expectation).

Reply:
> Here is a counter example for you

It's just an example for me, not a "counter example".
If `ok` parameter is purposely allowed to  be "omitted" (i.e. `nullptr`), then it is expected that function correctly works in both cases, either with `ok` exised or not. So this aspect should be tested.


> Note that the tests didn't cover it...

See. But after all, it is not covered by tests of the existing function `Utils::Gzip::compress (QByteArray, QByteArray, bool*)`, isn't it? So it became possible to make changes to the implementation containing such a bug, and this was missed by the tests.

But I definitely wouldn't expect passing null, uninitialized, not opened or opened in wrong mode `QIODevice` to be a valid use case. It can only exist if there is a logical error in the code. So it would be nice to first decide how to prevent such incorrect use, and only then do the tests. Judging by your statements, you propose to handle all errors (including logical ones) in the same way and approve this as expected behavior by the presence of appropriate tests. (At least it looks that way to me. If I'm wrong, just correct me.) IMO, it may make it harder to detect such logical errors if they appear in the future. Therefore, I would simply add assertions to all the preconditions, and I would cover by tests only the correctness of the main work of the function, provided that the parameters are acceptable.

Reply:
>It's just an example for me, not a "counter example".
If ok parameter is purposely allowed to be "omitted" (i.e. nullptr), then it is expected that function correctly works in both cases, either with ok exised or not. So this aspect should be tested.

To be clear, I was using the example to make a point that the interface should really be tested properly and it isn't "bring matters to the point of absurdity, when a function devotes 90% of logic to checking parameters instead of useful work." as you [claimed](https://github.com/qbittorrent/qBittorrent/pull/18645#discussion_r1243125742). Even if such thing exists it probably means that interface is really important and needs to be tested.

>But I definitely wouldn't expect passing null, uninitialized, not opened or opened in wrong mode QIODevice to be a valid use case. 

Why make this assumption? Under this (IMO inappropriate) assumption you won't catch errors that could happen in the real world. For example the device is valid for some time but just closed before calling into the tested function. Or it can even be closed when executing the function.

>Judging by your statements, you propose to handle all errors (including logical ones) in the same way and approve this as expected behavior by the presence of appropriate tests. 

I was thinking of using `std::expected` to report the errors when I wrote the reply.

>IMO, it may make it harder to detect such logical errors if they appear in the future. Therefore, I would simply add assertions to all the preconditions, and I would cover by tests only the correctness of the main work of the function, provided that the parameters are acceptable.

The concern of "harder to detect logical errors" is valid. But using assert only makes it untestable (and remember no test means no guarantee that it will be correct, you are relying on mere self confidence which means nothing). Throwing `std::logic_error` might be better since you can verify it with `QVERIFY_THROWS_EXCEPTION` and `QVERIFY_THROWS_NO_EXCEPTION`. Or perhaps `std::expected` can be utilized as well.

Reply:
> Why make this assumption?

In order to simplify the code. And this is a widely used approach used everywhere (do you really not understand what I'm talking about?). The only question is whether it is acceptable in this particular case or not. That is, what in this case is cheaper, more reliable, etc. - to do the processing of all the possible invalid parameters inside a function, or to respect some contract about acceptable parameters at the point of calling this function? I'm not sure exactly what this case is, so I keep debating about it.

Reply:
> Throwing `std::logic_error` might be better since you can verify it with `QVERIFY_THROWS_EXCEPTION` and `QVERIFY_THROWS_NO_EXCEPTION`. Or perhaps `std::expected` can be utilized instead.

At least I would strongly disapprove to combine both of them as error handling of the same function.

Reply:
> I was thinking of using `std::expected` to report the errors when I wrote the reply.

That's exactly what I thought. But it doesn't really matter which method of error handling is chosen. I feel it questionable to handle logical errors in the same way as normal runtime errors. Still, an error report like "writing to the file failed because the file has read-only attribute" makes sense for the user, but "nullptr is an invalid value for some parameter" - for the developer. Although, I may be mistaken, and this is quite normal.

Reply:
>The only question is whether it is acceptable in this particular case or not. That is, what in this case is cheaper, more reliable, etc. - to do the processing of all the possible invalid parameters inside a function, or to respect some contract about acceptable parameters at the point of calling this function? 

IMO if the function is local (reside in .cpp anonymous namespace or private function in some class) then it can be tailored, invalid parameters can be ignored or some preconditions/assertions can be made. However in our case here, the function is a bit more "global" so I would raise my expectation bar a bit higher. The (important) parameters should be properly checked and tested, since you cannot limit (or really know) which context it will reside.

>No, I'm talking specifically about the case when the function declaratively has "undefined" behavior in the case of invalid parameters. 

I don't understand. Why not try "define" it instead of leaving it "undefined"? BTW depending on the context, intentionally leaving behavior undefined *might* lead to a software vulnerability...

>At least I would strongly disapprove to combine both of them as error handling of the same function.

Agree.

>Still, an error report like "writing to the file failed because the file has read-only attribute" makes sense for the user, but "nullptr is an invalid value for some parameter" - for the developer. 

Perhaps the errors that are interesting for the users can be made into error status/messages, and other errors that are only interesting for the developers could be made as assertions (or exceptions)? In our case `QIODevice`, since a stable condition is not guaranteed (such as a unstable network drive). The function utilizing `QIODevice` should report the error to its caller (or user) when it happens IMO.

@brvphoenix I would suggest to leave this "error reporting" work for another time as to avoid bogging down this PR.